cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/utility/device_atomics.cuh>
+#include <cuspatial/utility/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+template <typename T, typename... Ts>
+constexpr bool is_same()
+{
+  return std::conjunction_v<std::is_same<T, Ts>...>;
+}
+
+template <typename... Ts>
+constexpr bool is_floating_point()
+{
+  return std::conjunction_v<std::is_floating_point<Ts>...>;
+}
```

Comment:
I like these utilities, and they would be useful in all of our APIs (I could have used them in haversine), so perhaps put them in type_utils.hpp?

Reply:
Since these are derived type traits, perhaps a better place is something like traits.hpp?